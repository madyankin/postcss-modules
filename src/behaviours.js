import localByDefault from "postcss-modules-local-by-default";
import extractImports from "postcss-modules-extract-imports";
import modulesScope from "postcss-modules-scope";
import values from "postcss-modules-values";

export const behaviours = {
  LOCAL: "local",
  GLOBAL: "global",
};

export function getDefaultPlugins({
  behaviour,
  generateScopedName,
  exportGlobals,
}) {
  const scope = modulesScope({ generateScopedName, exportGlobals });

  const plugins = {
    [behaviours.LOCAL]: [values, localByDefault({ mode: 'local' }), extractImports, scope],
    [behaviours.GLOBAL]: [values, localByDefault({ mode: 'global' }), extractImports, scope],
  };

  return plugins[behaviour];
}

export function isValidBehaviour(behaviour) {
  return (
    Object.keys(behaviours)
      .map((key) => behaviours[key])
      .indexOf(behaviour) > -1
  );
}
