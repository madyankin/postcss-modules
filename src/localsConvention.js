import camelCase from "lodash.camelcase";

function dashesCamelCase(string) {
	return string.replace(/-+(\w)/g, (_, firstLetter) => firstLetter.toUpperCase());
}

export function makeLocalsConventionReducer(localsConvention, inputFile) {
	const isFunc = typeof localsConvention === "function";

	return (tokens, [className, value]) => {
		if (isFunc) {
			const convention = localsConvention(className, value, inputFile);

			if (Array.isArray(convention)) {

				convention.forEach(convention => tokens[convention] = value);

				return tokens;

			}

			tokens[convention] = value;

			return tokens;
		}

		switch (localsConvention) {
			case "none":
				tokens[className] = value;
				break;

			case "all":
				tokens[className] = value;
				tokens[camelCase(className)] = value;
				tokens[dashesCamelCase(className)] = value;
				break;

			case "camelCase":
				tokens[className] = value;
				tokens[camelCase(className)] = value;
				break;

			case "camelCaseOnly":
				tokens[camelCase(className)] = value;
				break;

			case "dashes":
				tokens[className] = value;
				tokens[dashesCamelCase(className)] = value;
				break;

			case "dashesOnly":
				tokens[dashesCamelCase(className)] = value;
				break;
		}

		return tokens;
	};
}
