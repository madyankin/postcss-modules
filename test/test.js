import postcss from "postcss";
import autoprefixer from "autoprefixer";
import fs from "fs";
import path from "path";
import plugin from "../src";
import { behaviours } from "../src/behaviours";

const fixturesPath = path.resolve(__dirname, "./fixtures");

function createPlugin(name, processor) {
  const plugin = () => ({
    postcssPlugin: name,
    Once: processor,
  });
  plugin.postcss = true;
  return plugin;
}

const cases = {
  plugins: "saves origin plugins",
  classes: "processes classes",
  comments: "preserves comments",
  composes: "composes rules",
  values: "processes values",
  interpolated: "generates scoped name with interpolated string",
  global: "allows to make CSS global",
  localShorthand: "processes :local shorthand selector",
  globalShorthand: "processes :global shorthand selector",
};

function generateScopedName(name, filename) {
  const file = path.basename(filename, ".css").replace(/\./g, "_");
  return `_${file}_${name}`;
}

Object.keys(cases).forEach((name) => {
  const description = cases[name];

  const scopedNameGenerator =
    name === "interpolated"
      ? "[name]__[local]___[hash:base64:5]"
      : generateScopedName;

  const scopeBehaviour =
    name === behaviours.GLOBAL || name === "globalShorthand"
      ? behaviours.GLOBAL
      : behaviours.LOCAL;

  it(description, async () => {
    const sourceFile = path.join(fixturesPath, "in", `${name}.css`);
    const source = fs.readFileSync(sourceFile).toString();

    let resultJson;

    const plugins = [
      autoprefixer,
      plugin({
        scopeBehaviour,
        generateScopedName: scopedNameGenerator,
        getJSON: (cssFile, json) => {
          resultJson = json;
        },
      }),
    ];

    const result = await postcss(plugins).process(source, { from: sourceFile });

    expect(result.css).toMatchSnapshot(`${description} - CSS`);
    expect(resultJson).toMatchSnapshot(`${description} - JSON`);
  });

  it(`only calls plugins once when it ${description}`, async () => {
    const sourceFile = path.join(fixturesPath, "in", `${name}.css`);
    const source = fs.readFileSync(sourceFile).toString();

    const rootsSeenBeforePlugin = new Set();
    const rootsSeenAfterPlugin = new Set();

    const plugins = [
      autoprefixer,
      createPlugin("validator-1", (root) => {
        if (rootsSeenBeforePlugin.has(root)) {
          throw new Error("Plugin before ours was called multiple times.");
        }
        rootsSeenBeforePlugin.add(root);
        root.prepend(
          `/* validator-1-start (${path.basename(root.source.input.file)}) */`
        );
        root.append(
          `/* validator-1-end (${path.basename(root.source.input.file)}) */`
        );
      }),
      plugin({
        scopeBehaviour,
        generateScopedName: scopedNameGenerator,
        getJSON: () => {},
      }),
      createPlugin("validator-2", (root) => {
        if (rootsSeenAfterPlugin.has(root)) {
          throw new Error("Plugin after ours was called multiple times.");
        }
        rootsSeenAfterPlugin.add(root);
        root.prepend(
          `/* validator-2-start (${path.basename(root.source.input.file)}) */`
        );
        root.append(
          `/* validator-2-end (${path.basename(root.source.input.file)}) */`
        );
      }),
    ];

    const result = await postcss(plugins).process(source, { from: sourceFile });

    expect(result.css).toMatchSnapshot(`plugins once - ${description} - CSS`);
  });
});

it("works with visitor plugins", async () => {
  const source = `
p {
  color: green;
}
`;

  const plugins = [
    {
      postcssPlugin: "turn-values-blue",
      Declaration(decl) {
        decl.value = "blue";
      },
    },
    plugin(),
  ];
  const result = await postcss(plugins).process(source, { from: undefined });
  expect(result.css).toEqual(source.replace("green", "blue"));
});

it("saves JSON next to CSS by default", async () => {
  const sourceFile = path.join(fixturesPath, "in", "saveJSON.css");
  const source = fs.readFileSync(sourceFile).toString();
  const jsonFile = path.join(fixturesPath, "in", "saveJSON.css.json");

  if (fs.existsSync(jsonFile)) fs.unlinkSync(jsonFile);

  await postcss([plugin({ generateScopedName })]).process(source, {
    from: sourceFile,
  });

  const json = fs.readFileSync(jsonFile).toString();
  fs.unlinkSync(jsonFile);

  expect(JSON.parse(json)).toMatchObject({ title: "_saveJSON_title" });
});

it("processes globalModulePaths option", async () => {
  const sourceFile = path.join(fixturesPath, "in", "globalModulePaths.css");
  const source = fs.readFileSync(sourceFile).toString();

  const thePlugin = plugin({
    generateScopedName,
    globalModulePaths: [/globalModulePaths/],
    getJSON: () => {},
  });

  const result = await postcss([thePlugin]).process(source, {
    from: sourceFile,
  });

  expect(result.css).toMatchSnapshot("processes globalModulePaths option");
});

it("processes localsConvention with camelCase option", async () => {
  const sourceFile = path.join(fixturesPath, "in", "camelCase.css");
  const source = fs.readFileSync(sourceFile).toString();
  const jsonFile = path.join(fixturesPath, "in", "camelCase.css.json");

  if (fs.existsSync(jsonFile)) fs.unlinkSync(jsonFile);

  await postcss([
    plugin({ generateScopedName, localsConvention: "camelCase" }),
  ]).process(source, { from: sourceFile });

  const json = fs.readFileSync(jsonFile).toString();
  fs.unlinkSync(jsonFile);

  expect(JSON.parse(json)).toMatchObject({
    "camel-case": "_camelCase_camel-case",
    camelCase: "_camelCase_camel-case",
    "camel-case-extra": "_camelCase_camel-case-extra",
    camelCaseExtra: "_camelCase_camel-case-extra",
    FooBar: "_camelCase_FooBar",
    fooBar: "_camelCase_FooBar",
  });
});

it("processes localsConvention with camelCaseOnly option", async () => {
  const sourceFile = path.join(fixturesPath, "in", "camelCase.css");
  const source = fs.readFileSync(sourceFile).toString();
  const jsonFile = path.join(fixturesPath, "in", "camelCase.css.json");

  if (fs.existsSync(jsonFile)) fs.unlinkSync(jsonFile);

  await postcss([
    plugin({ generateScopedName, localsConvention: "camelCaseOnly" }),
  ]).process(source, { from: sourceFile });

  const json = fs.readFileSync(jsonFile).toString();
  fs.unlinkSync(jsonFile);

  expect(JSON.parse(json)).toMatchObject({
    camelCase: "_camelCase_camel-case",
    camelCaseExtra: "_camelCase_camel-case-extra",
    fooBar: "_camelCase_FooBar",
  });
});

it("processes localsConvention with dashes option", async () => {
  const sourceFile = path.join(fixturesPath, "in", "camelCase.css");
  const source = fs.readFileSync(sourceFile).toString();
  const jsonFile = path.join(fixturesPath, "in", "camelCase.css.json");

  if (fs.existsSync(jsonFile)) fs.unlinkSync(jsonFile);

  await postcss([
    plugin({ generateScopedName, localsConvention: "dashes" }),
  ]).process(source, { from: sourceFile });

  const json = fs.readFileSync(jsonFile).toString();
  fs.unlinkSync(jsonFile);

  expect(JSON.parse(json)).toMatchObject({
    "camel-case": "_camelCase_camel-case",
    camelCase: "_camelCase_camel-case",
    "camel-case-extra": "_camelCase_camel-case-extra",
    camelCaseExtra: "_camelCase_camel-case-extra",
    FooBar: "_camelCase_FooBar",
  });
});

it("processes localsConvention with dashes option", async () => {
  const sourceFile = path.join(fixturesPath, "in", "camelCase.css");
  const source = fs.readFileSync(sourceFile).toString();
  const jsonFile = path.join(fixturesPath, "in", "camelCase.css.json");

  if (fs.existsSync(jsonFile)) fs.unlinkSync(jsonFile);

  await postcss([
    plugin({ generateScopedName, localsConvention: "dashes" }),
  ]).process(source, { from: sourceFile });

  const json = fs.readFileSync(jsonFile).toString();
  fs.unlinkSync(jsonFile);

  expect(JSON.parse(json)).toMatchObject({
    camelCase: "_camelCase_camel-case",
    camelCaseExtra: "_camelCase_camel-case-extra",
    FooBar: "_camelCase_FooBar",
  });
});

it("processes localsConvention with function option", async () => {
  const sourceFile = path.join(fixturesPath, "in", "camelCase.css");
  const source = fs.readFileSync(sourceFile).toString();
  const jsonFile = path.join(fixturesPath, "in", "camelCase.css.json");

  if (fs.existsSync(jsonFile)) fs.unlinkSync(jsonFile);

  await postcss([
    plugin({
      generateScopedName,
      localsConvention: (className) => {
        return className.replace("camel-case", "cc");
      },
    }),
  ]).process(source, { from: sourceFile });

  const json = fs.readFileSync(jsonFile).toString();
  fs.unlinkSync(jsonFile);

  expect(JSON.parse(json)).toMatchObject({
    cc: "_camelCase_camel-case",
    "cc-extra": "_camelCase_camel-case-extra",
    FooBar: "_camelCase_FooBar",
  });
});

it("processes hashPrefix option", async () => {
  const generateScopedName = "[hash:base64:5]";
  const hashPrefix = "prefix";
  const getJSON = () => {};

  const withoutHashPrefix = plugin({ generateScopedName, getJSON });
  const withHashPrefix = plugin({ generateScopedName, getJSON, hashPrefix });

  const css = ".foo {}";
  const params = { from: "test.css" };

  const result1 = await postcss([withoutHashPrefix]).process(css, params);
  const result2 = await postcss([withHashPrefix]).process(css, params);

  expect(result2.css).toMatchSnapshot("processes hashPrefix option");
  expect(result1.css).not.toEqual(result2.css);
});

it("different instances have different generateScopedName functions", async () => {
  const one = plugin({
    generateScopedName: () => "one",
    getJSON: () => {},
  });

  const two = plugin({
    generateScopedName: () => "two",
    getJSON: () => {},
  });

  const css = ".foo {}";
  const params = { from: "test.css" };

  const resultOne = await postcss([one]).process(css, params);
  const resultTwo = await postcss([two]).process(css, params);

  expect(resultOne.css).toEqual(".one {}");
  expect(resultTwo.css).toEqual(".two {}");
});

it("getJSON with outputFileName", async () => {
  const sourceFile = path.join(fixturesPath, "in", "test", "getJSON.css");
  const expectedFile = path.join(fixturesPath, "out", "test", "getJSON");
  const source = fs.readFileSync(sourceFile).toString();
  const expectedJSON = fs.readFileSync(`${expectedFile}.json`).toString();
  let jsonFileName;
  let resultJson;

  const plugins = [
    plugin({
      generateScopedName,
      getJSON: (cssFile, json, outputFileName) => {
        jsonFileName = outputFileName.replace(".css", ".json");
        resultJson = json;
      },
    }),
  ];

  await postcss(plugins).process(source, {
    from: sourceFile,
    to: `${expectedFile}.css`,
  });

  expect(jsonFileName).toEqual(`${expectedFile}.json`);
  expect(resultJson).toMatchObject(JSON.parse(expectedJSON));
});

it("exposes export tokens for other plugins", async () => {
  const sourceFile = path.join(fixturesPath, "in", "values.css");
  const source = fs.readFileSync(sourceFile).toString();

  const plugins = [
    plugin({
      generateScopedName,
      getJSON: () => {},
    }),
  ];

  const result = await postcss(plugins).process(source, {
    from: sourceFile,
  });

  expect(result.messages).toMatchSnapshot(
    "exposes export tokens for other plugins"
  );
});

it("processes exportGlobals option", async () => {
  const sourceFile = path.join(fixturesPath, "in", "classes.css");
  const source = fs.readFileSync(sourceFile).toString();
  let json;

  await postcss([
    plugin({
      generateScopedName,
      exportGlobals: true,
      getJSON: (_, result) => {
        json = result;
      },
    }),
  ]).process(source, { from: sourceFile });

  expect(json).toMatchObject({
    page: "page",
    title: "_classes_title",
    article: "_classes_article",
  });
});

it("processes resolve option", async () => {
  const sourceFile = path.join(fixturesPath, "in", "compose.resolve.css");
  const source = fs.readFileSync(sourceFile).toString();
  let json;
  const result = await postcss([
    plugin({
      generateScopedName,
      resolve: async (file) => {
        return file.replace(/^test-fixture-in/, path.dirname(sourceFile));
      },
      getJSON: (_, result) => {
        json = result;
      },
    }),
  ]).process(source, { from: sourceFile });

  expect(result.css).toMatchSnapshot("processes resolve option");
  expect(json).toStrictEqual({
    figure: "_compose_resolve_figure _composes_a_hello",
    "figure-single-quote":
      "_compose_resolve_figure-single-quote _composes_a_hello",
  });
});
