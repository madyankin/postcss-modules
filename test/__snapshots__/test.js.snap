// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows to make CSS global: allows to make CSS global - CSS 1`] = `
".page {
    padding: 20px;
}

._global_title {
    color: green;
}
"
`;

exports[`allows to make CSS global: allows to make CSS global - JSON 1`] = `
Object {
  "title": "_global_title",
}
`;

exports[`composes rules: composes rules - CSS 1`] = `
"._composes_a_another-mixin {
    display: -webkit-flex;
    display: -moz-box;
    display: flex;
    height: 100px;
    width: 200px;
}._composes_a_hello {
    foo: bar;
}._composes_mixins_title {
    color: black;
    font-size: 40px;
}._composes_mixins_title:hover {
    color: red;
}._composes_mixins_figure {
    text-align: center
}._composes_mixins_title:focus, ._composes_mixins_figure:focus {
    outline: none;
    border: 1px solid red;
}

.page {
    padding: 20px;
}

._composes_title {
    color: green;
}

._composes_article {
    font-size: 16px;
}

._composes_figure {
    display: -webkit-flex;
    display: -moz-box;
    display: flex;
}
"
`;

exports[`composes rules: composes rules - JSON 1`] = `
Object {
  "article": "_composes_article",
  "figure": "_composes_figure _composes_mixins_figure _composes_a_another-mixin",
  "title": "_composes_title _composes_mixins_title",
}
`;

exports[`exposes export tokens for other plugins: exposes export tokens for other plugins 1`] = `
Array [
  Object {
    "exportTokens": Object {
      "article": "_values_article",
      "colors": "\\"./values.colors.css\\"",
      "primary": "green",
      "secondary": "blue",
      "title": "_values_title",
    },
    "plugin": "postcss-modules",
    "type": "export",
  },
]
`;

exports[`generates scoped name with interpolated string: generates scoped name with interpolated string - CSS 1`] = `
".interpolated__title___mAhtv {
    color: green;
}

.interpolated__article___MQq7x {
    color: black;
}
"
`;

exports[`generates scoped name with interpolated string: generates scoped name with interpolated string - JSON 1`] = `
Object {
  "article": "interpolated__article___MQq7x",
  "title": "interpolated__title___mAhtv",
}
`;

exports[`only calls plugins once when it allows to make CSS global: plugins once - allows to make CSS global - CSS 1`] = `
"/* validator-2-start (global.css) */

/* validator-1-start (global.css) */

.page {
    padding: 20px;
}

._global_title {
    color: green;
}

/* validator-1-end (global.css) */

/* validator-2-end (global.css) */
"
`;

exports[`only calls plugins once when it composes rules: plugins once - composes rules - CSS 1`] = `
"/* validator-1-start (composes.a.css) */._composes_a_another-mixin {
    display: -webkit-flex;
    display: -moz-box;
    display: flex;
    height: 100px;
    width: 200px;
}._composes_a_hello {
    foo: bar;
}/* validator-1-end (composes.a.css) *//* validator-1-start (composes.mixins.css) */._composes_mixins_title {
    color: black;
    font-size: 40px;
}._composes_mixins_title:hover {
    color: red;
}._composes_mixins_figure {
    text-align: center
}._composes_mixins_title:focus, ._composes_mixins_figure:focus {
    outline: none;
    border: 1px solid red;
}/* validator-1-end (composes.mixins.css) */

/* validator-2-start (composes.css) */

/* validator-1-start (composes.css) */

.page {
    padding: 20px;
}

._composes_title {
    color: green;
}

._composes_article {
    font-size: 16px;
}

._composes_figure {
    display: -webkit-flex;
    display: -moz-box;
    display: flex;
}

/* validator-1-end (composes.css) */

/* validator-2-end (composes.css) */
"
`;

exports[`only calls plugins once when it generates scoped name with interpolated string: plugins once - generates scoped name with interpolated string - CSS 1`] = `
"/* validator-2-start (interpolated.css) */

/* validator-1-start (interpolated.css) */

.interpolated__title___mAhtv {
    color: green;
}

.interpolated__article___MQq7x {
    color: black;
}

/* validator-1-end (interpolated.css) */

/* validator-2-end (interpolated.css) */
"
`;

exports[`only calls plugins once when it preserves comments: plugins once - preserves comments - CSS 1`] = `
"/* validator-2-start (comments.css) */
/* validator-1-start (comments.css) */
/**
 * This is a doc comment...
 */
p {
  /* ...and a line comment */
  padding: 0;
}
/* validator-1-end (comments.css) */
/* validator-2-end (comments.css) */
"
`;

exports[`only calls plugins once when it processes :global shorthand selector: plugins once - processes :global shorthand selector - CSS 1`] = `
"/* validator-2-start (globalShorthand.css) *//* validator-1-start (globalShorthand.css) */
.title {
    color: red;
}/* validator-1-end (globalShorthand.css) *//* validator-2-end (globalShorthand.css) */
"
`;

exports[`only calls plugins once when it processes :local shorthand selector: plugins once - processes :local shorthand selector - CSS 1`] = `
"/* validator-2-start (localShorthand.css) *//* validator-1-start (localShorthand.css) */
._localShorthand_title {
    color: red;
}/* validator-1-end (localShorthand.css) *//* validator-2-end (localShorthand.css) */
"
`;

exports[`only calls plugins once when it processes classes: plugins once - processes classes - CSS 1`] = `
"/* validator-2-start (classes.css) */

/* validator-1-start (classes.css) */

.page {
    padding: 20px;
}

._classes_title {
    color: green;
}

._classes_article {
    color: black;
}

/* validator-1-end (classes.css) */

/* validator-2-end (classes.css) */
"
`;

exports[`only calls plugins once when it processes values: plugins once - processes values - CSS 1`] = `
"/* validator-1-start (values.colors.css) *//* validator-1-end (values.colors.css) */
/* validator-2-start (values.css) */
/* validator-1-start (values.css) */

._values_title {
    color: green;
}

._values_article {
    color: blue;
}

/* validator-1-end (values.css) */

/* validator-2-end (values.css) */
"
`;

exports[`only calls plugins once when it saves origin plugins: plugins once - saves origin plugins - CSS 1`] = `
"/* validator-2-start (plugins.css) *//* validator-1-start (plugins.css) */
._plugins_title {
    display: -webkit-flex;
    display: -moz-box;
    display: flex;
    color: green;
}/* validator-1-end (plugins.css) *//* validator-2-end (plugins.css) */
"
`;

exports[`preserves comments: preserves comments - CSS 1`] = `
"/**
 * This is a doc comment...
 */
p {
  /* ...and a line comment */
  padding: 0;
}
"
`;

exports[`preserves comments: preserves comments - JSON 1`] = `Object {}`;

exports[`processes :global shorthand selector: processes :global shorthand selector - CSS 1`] = `
".title {
    color: red;
}
"
`;

exports[`processes :global shorthand selector: processes :global shorthand selector - JSON 1`] = `Object {}`;

exports[`processes :local shorthand selector: processes :local shorthand selector - CSS 1`] = `
"._localShorthand_title {
    color: red;
}
"
`;

exports[`processes :local shorthand selector: processes :local shorthand selector - JSON 1`] = `
Object {
  "title": "_localShorthand_title",
}
`;

exports[`processes classes: processes classes - CSS 1`] = `
".page {
    padding: 20px;
}

._classes_title {
    color: green;
}

._classes_article {
    color: black;
}
"
`;

exports[`processes classes: processes classes - JSON 1`] = `
Object {
  "article": "_classes_article",
  "title": "_classes_title",
}
`;

exports[`processes globalModulePaths option: processes globalModulePaths option 1`] = `
".page {
    padding: 20px;
}

._globalModulePaths_title {
    color: green;
}
"
`;

exports[`processes hashPrefix option: processes hashPrefix option 1`] = `".uG5tz {}"`;

exports[`processes resolve option: processes resolve option 1`] = `
"._composes_a_another-mixin {
    display: flex;
    height: 100px;
    width: 200px;
}._composes_a_hello {
    foo: bar;
}

._compose_resolve_figure {
    display: flex;
}

._compose_resolve_figure-single-quote {
    display: block;
}
"
`;

exports[`processes values: processes values - CSS 1`] = `
"

._values_title {
    color: green;
}

._values_article {
    color: blue;
}
"
`;

exports[`processes values: processes values - JSON 1`] = `
Object {
  "article": "_values_article",
  "colors": "\\"./values.colors.css\\"",
  "primary": "green",
  "secondary": "blue",
  "title": "_values_title",
}
`;

exports[`saves origin plugins: saves origin plugins - CSS 1`] = `
"._plugins_title {
    display: -webkit-flex;
    display: -moz-box;
    display: flex;
    color: green;
}
"
`;

exports[`saves origin plugins: saves origin plugins - JSON 1`] = `
Object {
  "title": "_plugins_title",
}
`;
